[Post] Utilising API Blueprint in API clients (Swift)

kylef opened the pull request:
- [x] Initial review 
- [x] Fix code-examples (they need to be highlighted correctly)
- [x] Release representor 0.2.0
- [x] Update link to https://github.com/apiaryio/api-blueprint/pull/148 once merged
- [x] [Release code examples](https://github.com/apiaryio/apiary/pull/2158)
- [x] Include image of code examples
- [x] Final review

zdne commented:
@kylef I would actually like to see a bigger example, given maybe referring to the basic tutorial blueprint in the combination how a client should work + education more in-depth on some of the benefits (and showing possible transition towards siren / hal). 

What do you think? 

Also @smizell or @fosrias could be interested in this and help us to sell representor ideas ...

smizell commented on _posts/2015-02-04-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,57 @@
+
+Imagine an API client when you specify which resource and action you want to perform from an API Blueprint, it can use the already defined information from a blueprint to build the request without having to duplicate the information such as the HTTP method and URI from your blueprint. This allows you to start building smarter clients which can evolve at run-time based on the blueprint.
+
+Let’s walk though how this looks in Swift with an example. We’re going to start my loading our blueprint file that’s embedded in our application. You’re not limited to embedding it in your application, you could even retrieve it dynamically.
```
"We’re going to start **by** loading"

smizell replied:
> You’re not limited to embedding it in your application, you could even retrieve it dynamically.

I think this should be two sentence, or joined with "and."

smizell commented on _posts/2015-02-04-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,57 @@
+
+```swift
+"We can perform the transition with the URI \(transition.uri) using the HTTP method \(transition.method)."
+```
````
Is this a Swift code block?

kylef replied:
Yes, `\()` is for string interpolation.

kylef commented:
@zdne I've updated this pull request to try and address your feedback. Please take a look at let me know if I went in the right direction.

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+Imagine an API client when you specify which resource and action you want to perform from an API Blueprint, it can use the already defined information from a blueprint to build the request without having to duplicate the information such as the HTTP method and URI from your blueprint. This allows you to start building smarter clients which can evolve at run-time based on the blueprint.
+
+The SDK is built on-top of a framework we’ve built called the [Representor](https://github.com/the-hypermedia-project/charter), which is designed to be a canonical representation of a hypermedia message in multiple languages. One of the key constraints to Representational State Transfer (REST) is Hypermedia as the Engine of State (HATEOS).  The message from the server itself should contain inform ...
```
Its HATEOAS  (not HATEOS)

fosrias replied:
How about:

The SDK is built on-top of a framework we’ve built that implements the [Representor Pattern](https://github.com/the-hypermedia-project/charter), which provides canonical representations of a hypermedia message in multiple languages.

Or something like that.

fosrias replied:
Hypermedia as the Engine of Application State (HATEOAS)

smizell replied:
To add comments to the suggestion by @fosrias

- Maybe do not use "built" twice in the sentence?
- Multiple media types instead of multiple languages?

smizell replied:
Also, last sentence could be tied the previous more with:

"This constraints requires the message from the server to contain information about how to transition to other states along with links to other resources, specifically through hypermedia."

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+Since a blueprint is effectively a document which specifies all of the transitions (actions) in your API, we’ve built on-top of the Representor and added the ability to deserialise an API Blueprint into a collection of transitions.
+
+This allows us to decouple an API client from the server-side implementation, we no longer have to know about the URIs to each resource and the HTTP methods in our code. We can negotiate these at run-time, similar to how with HATEOS we follow links, we can follow links from an API Blueprint.
```
HATEOAS

smizell replied:
After "implementation," I think that sentence should end.

Last sentence could read, since following links is HATEOAS (IMO):

We can negotiate these at run-time and follow links from an API Blueprint.

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+Once you’ve loaded your blueprint, you can retrieve the different actions as [transitions](https://github.com/the-hypermedia-project/charter/blob/master/reference/hypermedia-elements.md#transitions) from your blueprint.
+
+This example is going to make use of one of our example blueprints, “Polls API”. I’ve pulled out an excerpt of the example blueprint and it looks something like the following. You can find the full version over in our [examples](https://github.com/apiaryio/api-blueprint/blob/kylef/polls-example/examples/Polls%20API.md).
```
I think we should point this instead of GH to Apiary. Create a new API under Apiary team and copy paste there that example (and yes I know :((), then don't forget to add the link to GH example too (what if somebody wants to see the source). 

And yes, this is making me really sad – years I am trying to get the access to API blueprints but we still have to go through this.

smizell replied:
"We will use our sample "Polls API" blueprint for this example."

I would also be tempted to hyperlink "full version" as you are not linking to examples, but the full version.

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+
+This example is going to make use of one of our example blueprints, “Polls API”. I’ve pulled out an excerpt of the example blueprint and it looks something like the following. You can find the full version over in our [examples](https://github.com/apiaryio/api-blueprint/blob/kylef/polls-example/examples/Polls%20API.md).
+
+```markdown
````
Also wouldn't be cool if we could actually embed our own doc here ? cc @janmoravec

kylef replied:
That's a great idea. We could make use of oEmbed like Vimeo/YouTube do with iframes: http://www.oembed.com.

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+We can use the elegant Swift HTTP networking library [Alamofire](https://github.com/Alamofire/Alamofire) along with some convenience methods we’ve built to perform HTTP requests from our transition directly.
+
+```swift
+let baseURL = NSURL(string: "http://polls.apiblueprint.org/")
````
Out of curiosity – isn't that root URL hidden somewhere in the blueprint Adapter?

zdne commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+If you’re interested in giving this a go, check out our [installation guide](https://github.com/the-hypermedia-project/representor-swift/wiki/API-Blueprint) on using API Blueprint in Swift.
+
+Of course, we’ve integrated this into our existing code examples. So right from Apiary you can see how it looks to perform a transition in Swift.
```
Looks good I would just add some call to action to the end (such as let us know what do you think or something like that).

smizell replied:
What?! It's already in the code examples. Very cool!

zdne commented:
Reviewed. Looks good to me. Also note @fosrias and @smizell this is really the public launch of Hypermedia Project. Just sayin'

fosrias commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+The SDK is built on-top of a framework we’ve built called the [Representor](https://github.com/the-hypermedia-project/charter), which is designed to be a canonical representation of a hypermedia message in multiple languages. One of the key constraints to Representational State Transfer (REST) is Hypermedia as the Engine of State (HATEOS).  The message from the server itself should contain inform ...
+
+The Representor is a structure providing this functionality along with the ability to deserialise Hypermedia formats such as [HAL](http://stateless.co/hal_specification.html) and [Siren](https://github.com/kevinswiber/siren). The Representer contains a couple of properties, one of them is it’s transitions to other states, along with how the current resource links to other resources.
```
This will need a little massaging to define a Representor as an implementation of the pattern in swift, but other languages coming or something.

fosrias replied:
This is based on my prior comment about using the Representor Pattern

fosrias commented:
@kylef @zdne Just thinking out loud here, but we talked about a sample application at the app store around this. Does that work with the polls API? Would we want to OSS that code and also link it here (and maybe hold off going public a bit to pull that in)?

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+comments: true
+---
+
+I’m excited to release a new way to interface to API Blueprints. A client-side framework in Swift that allows you to introspect and perform actions at run-time.
```
The last sentence here is a sentence fragment. Maybe consider how to tie these two sentences together.

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+I’m excited to release a new way to interface to API Blueprints. A client-side framework in Swift that allows you to introspect and perform actions at run-time.
+
+Imagine an API client when you specify which resource and action you want to perform from an API Blueprint, it can use the already defined information from a blueprint to build the request without having to duplicate the information such as the HTTP method and URI from your blueprint. This allows you to start building smarter clients which can evolve at run-time based on the blueprint.
```
You could probably simplify this first sentence, as it was somewhat tautological:

"Imagine an API client that can use the already-defined information from an API Blueprint to build a request rather than hardcoding that information. This allows you to start building smarter clients which can evolve at run-time based on the blueprint."

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,119 @@
+
+The Representor is a structure providing this functionality along with the ability to deserialise Hypermedia formats such as [HAL](http://stateless.co/hal_specification.html) and [Siren](https://github.com/kevinswiber/siren). The Representer contains a couple of properties, one of them is it’s transitions to other states, along with how the current resource links to other resources.
+
+Since a blueprint is effectively a document which specifies all of the transitions (actions) in your API, we’ve built on-top of the Representor and added the ability to deserialise an API Blueprint into a collection of transitions.
```
Is this this correct that a blueprint specifies all transitions?

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+
+```swift
+let blueprint = Blueprint(named: "blueprint.json", bundle: nil)
+```
````
Does it require to use the AST? (because of the .json)

kylef replied:
It does require the AST. I've linked to an installation guide near the bottom of the post which mentions how to use Snowcrash to grab the AST: https://github.com/the-hypermedia-project/representor-swift/wiki/API-Blueprint.

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+let blueprint = Blueprint(named: "blueprint.json", bundle: nil)
+```
+
+Once you’ve loaded your blueprint, you can retrieve the different actions as [transitions](https://github.com/the-hypermedia-project/charter/blob/master/reference/hypermedia-elements.md#transitions) from your blueprint.
````
- Maybe "different **resource** actions"
- You can remove "from your blueprint."

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+This transition allows us to introspect it’s values at run-time, allowing us to retrieve protocol information such as the HTTP method and URI along with all the URI parameters.
````
```diff
- it's
+ its
```

smizell commented on _posts/2015-02-06-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+Which will result in performing a PUT request to `/questions/1/choices/2`.
````
IMO

```diff
- Which
+ This
```

smizell commented:
@kylef I read through for grammar and clarity. It all looks great! I think hitting more on what a Representor is per the comments of @fosrias would be good, too, though maybe that could be another blog post to go into deeper detail?

zdne commented:
I concur with @smizell and @fosrias comments. Also

>  though maybe that could be another blog post to go into deeper detail?

Is something we  (@fosrias) want to do as well?

fosrias commented:
@zdne I am game for another blog post in more detail.

kylef commented:
I think I've addressed all of your feedback, please have another look over the post @fosrias @smizell @zdne.

> @kylef @zdne Just thinking out loud here, but we talked about a sample application at the app store around this. Does that work with the polls API? Would we want to OSS that code and also link it here (and maybe hold off going public a bit to pull that in)?

I like what you're thinking here, perhaps this can be mentioned in the blog post announcing the new API Blueprint tutorials instead?

fosrias commented on _posts/2015-02-16-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,120 @@
+
+Imagine an API client that can use the already-defined information from an API Blueprint to build a request rather than hardcoding that information. This allows you to start building smarter clients which can evolve at run-time based on the blueprint.
+
+The SDK is created using a framework we’ve built which implements the [Representor Pattern](https://github.com/the-hypermedia-project/charter), a canonical representation of hypermedia messages across multiple media types and languages. One of the key constraints to Representational State Transfer (REST) is Hypermedia as the Engine of Application State (HATEOAS). This constraint requires the mess ...
```
Should this be: https://github.com/the-hypermedia-project/charter#representor-pattern

fosrias commented on _posts/2015-02-16-Utilising-API-Blueprint-in-API-Clients.md:
````diff
@@ -0,0 +1,120 @@
+We can pull out any of the actions from our blueprint by the resource and action name in Swift:
+
+```swift
+if let transition = blueprint?.transition("Choice", action: "Vote") {
````
This looks funny when rendered. Maybe add a comment like:

```swift
if let transition = blueprint?.transition("Choice", action: "Vote") {
  // Do something.
}
```

fosrias commented on _posts/2015-02-16-Utilising-API-Blueprint-in-API-Clients.md:
```diff
@@ -0,0 +1,120 @@
+
+<img src="/images/2015-02-16-Utilising-API-Blueprint-in-API-Clients/code-example.png" width="665" height="364" alt="API Blueprint Representor Code Examples" />
+
+We would love to hear your thoughts in the comments section below.
```
The only thing I think missing below is too put and else block with a comment about what you would do to recover or alternate path for loose coupling. E.g.

```swift
if transition //...
  // your current implementation
else
  // Gracefully handle the absence of the transition (or something like this)
end
```

fosrias commented:
@kylef Looks great. Like the tone. Added a few comments.

zdne commented:
@kylef please and or react on Mark's remarks, thanks.

kylef commented:
All addressed and updated the published date to today.

zdne commented:
Re-reading

zdne commented:
I do not have any further comments. Let's do it after https://github.com/apiaryio/api-blueprint/pull/148 is merged

zdne commented:
https://github.com/apiaryio/api-blueprint/pull/148 merged, @kylef update the link in the article?

kylef commented:
@zdne Updated, this is good to be merged / released now.

zdne commented:
@kylef Good, once @Baggz deploys the code examples I will release the blog post! Thanks everyone. Good job! :balloon:
